File: Data Structures & Algorithms/find-leaves-of-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<vector<int>> res;
    int dfs(TreeNode* root){
        if (!root) return -1;
        int idx = max(dfs(root->left), dfs(root->right))+1;
        vector<int> v = {root->val};
        idx == res.size() ? res.push_back(v) : res[idx].push_back(root->val);
        return idx;
    }

    vector<vector<int>> findLeaves(TreeNode* root) {
        dfs(root);
        return res;
    }
};

File: Data Structures & Algorithms/lowest-common-ancestor-of-a-binary-tree/submission-0.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
class Solution {
    /**
     * @param {TreeNode} root
     * @param {TreeNode} p
     * @param {TreeNode} q
     * @return {TreeNode}
     */
    dfs(root, p, q) {
        if (root == p) return p;
        if (root == q) return q;
        if (!root) return;

        let leftSearch = this.dfs(root.left, p, q);
        let rightSearch = this.dfs(root.right, p, q);

        if (rightSearch && leftSearch) return root;
        if (!rightSearch && leftSearch) return leftSearch;
        if (rightSearch && !leftSearch) return rightSearch;
    }
    
    lowestCommonAncestor(root, p, q) {
        return this.dfs(root, p, q);
    }
}
